docs/proposals/scheduler/README.md

```diff
@@ -0,0 +1,130 @@
+* `api/handler`: Handle API calls and redirect them properly
+* `node/watcher`: Running continuous watch process to know when a node is down
+* `cluster/monitor`: Handle API calls relative to statistics and nodes monitoring
+* `cluster/state_save`: Save the current state of the cluster 
+
+## Communication in the cluster 
+
+There are mainly two solutions. Either having direct connections between components (HTTP, gRPC...) of the cluster or having
+message queues where information is centralized.
+
+The second solution can be a great improvement for scalability and performance as everything would be completely asynchronous. 
+However, we need synchronous communication between internal services so make sure every information is properly received & understood.
+That's why the first solution direct communication between major components is a good choice. The drawback of this solution is we have 
+to write APIs for each component. 
+
+On top of that, we will be using [gRPC](https://grpc.io/) for communication between components. It will be handy to use 
+as API definitions are defined through [protoBuf](https://developers.google.com/protocol-buffers). 
+
+There is an interrogation around the controller & scheduler communication. They may be in the same 
+physical machine and not isolated one from another, so do we still need to use gRPC here ? Can't we 
+use any other solution of communication, as we are on the same physical machine ?
+
+The APIs exposed by our components must be defined through the need defined by the team 
+[controller](#controller) and the team [node](#node)
+
+---
+
+
+## Watcher 
+
+### Events 
+
+Watcher will work with an etcd to store usual and non highly dynamical data.
+The etcd will store:
+ - Number of workers
+ - workers alias (to improve/simplify communication)
+ - worker properties (such as cpu, RAM, memory, etc)
+
+Watcher is here to handle worker data, requesting it to the node manager/agent throught an API.
+Watcher will store in RAM the actual state of each nodes (idle, running, reloading, crashing, etc) and can give this metrics to scheduler when needed. If a node crash/restart, watcher will update datas in etcd.
+Watcher is the only point of communication with workers, it mean that he have to transfert scheduler instructions to the appropriate worker by resolving it using etcd alias name.
+
+### Recovery in case of crash
+
+If the watcher die or crash, he will restart by himself or launched by the scheduler.
+At loading, it ask in etcd all workers location, and for each, ask their actual state throught API, so all required data to work will be recovered.
+
+--- 
+
+The following parts explains everything needed from other teams.
+
+## Controller
+
+* What are the informations and events the controller needs to know about the cluster ?
+
+## Node 
+
+* What is the sending frequency of informations ? Does it need to be dynamic, if so we need the controller to know that 
+as he will manage this state. 
```

Comment:
For the moment, there is not frequency notion in our metrics behavior. 

First, I think we have to define how our components communicate with each other. By knowing that, we can discuss of how the metrics will be collected. (And all the resources) because the format of payload can be different between protocols.

By the way, I maintain my first idea, the watcher have to call our metrics endpoint frequently. It's quite simple, a CRON job should be enough. You can take example of Kubernetes metrics-server.

Reply:
Understandable., no problem on putting this question in stand by. 

About the watcher calling nodes endpoints frequently, it's part of the idea. Based on our discussions and your schema ([ref.](https://github.com/thomasgouveia/rik/tree/node-agent-enhancement/docs/rep/node-agent)), we will either wait for your data (you send data to watcher) and call for "ping" (what the call really does need to be defined) frequently to know whether or not your worker node is up.